tika/server.go

```diff
@@ -0,0 +1,230 @@
+// for the Server to start (default 10s).
+func WithStartupTimeout(d time.Duration) Option {
+	return func(s *Server) {
+		s.startupTimeout = d
+	}
+}
+
+// NewServer creates a new Server.
+func NewServer(jar string, options ...Option) (*Server, error) {
+	if jar == "" {
+		return nil, fmt.Errorf("no jar file specified")
+	}
+	if _, err := os.Stat(jar); err != nil {
+		return nil, fmt.Errorf("jar file not found: %s", jar)
+	}
+	s := &Server{
+		jar:            jar,
+		port:           "9998",
+		startupTimeout: 10 * time.Second,
+		hostname:       "localhost",
+	}
+	for _, o := range options {
+		o(s)
+	}
+	urlString := "http://" + s.hostname + ":" + s.port
+	u, err := url.Parse(urlString)
+	if err != nil {
+		return nil, fmt.Errorf("invalid hostname %q or port %q: %v", s.hostname, s.port, err)
+	}
+	s.url = u.String()
+	return s, nil
+}
+
+type commander func(context.Context, string, ...string) *exec.Cmd
+
+// cmder is used to stub out *exec.Cmd for testing.
+var cmder commander = exec.CommandContext
+
+// Start starts the given server. Start will start a new Java process. The
+// caller must call cancel() to shut down the process when finished with the
+// Server. The given Context is used for the Java process, not for cancellation
+// of startup.
+func (s *Server) Start(ctx context.Context) (cancel func(), err error) {
+	ctx, cancel = context.WithCancel(ctx)
+	cmd := cmder(ctx, "java", "-jar", s.jar, "-p", s.port)
+
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	if err := cmd.Start(); err != nil {
+		cancel()
+		return nil, err
+	}
+
+	if err := s.waitForStart(ctx); err != nil {
+		cancel()
+		buf, readErr := ioutil.ReadAll(stderr)
+		if readErr != nil {
+			return nil, fmt.Errorf("error reading stderr: %v", readErr)
+		}
+		// Report stderr since sometimes the server says why it failed to start.
+		return nil, fmt.Errorf("error starting server: %v\nserver stderr:\n\n%v", err, string(buf))
+	}
+	return cancel, nil
+}
+
+// waitForServer waits until the given Server is responding to requests.
+// waitForStart returns an error if the server does not respond within the
+// timeout set by WithStartupTimeout or if ctx is Done() first.
+func (s Server) waitForStart(ctx context.Context) error {
+	c := NewClient(nil, s.url)
+	ctx, cancel := context.WithTimeout(ctx, s.startupTimeout)
+	defer cancel()
+	for {
+		select {
+		case <-time.Tick(500 * time.Millisecond):
+			if _, err := c.Version(ctx); err == nil {
+				return nil
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func validateFileMD5(path, wantH string) (bool, string) {
+	f, err := os.Open(path)
+	if err != nil {
+		return false, ""
+	}
+	defer f.Close()
+
+	h := md5.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return false, ""
+	}
+	md5 := fmt.Sprintf("%x", h.Sum(nil))
+	return md5 == wantH, md5
+}
+
+// A Version represents a Tika Server version.
+type Version string
```

Comment:
Why is this type exported?

Reply:
It doesn't need to be. I made it unexported.